### DISCUSSION TITLE
Draw line 1 pixel width

### DISCUSSION THREAD

**Discussion by albertofll** [AngusJohnson/Image32#21] [General]
Hello I begin to playing with this library to check the possibility of using it for a drawing and text editor. The come from GDI and drawing a line with this code:

`img.AntiAliased:= false;

  path2[0].X:= 50;
  path2[0].Y:= 150;

  path2[1].X:= 400;
  path2[1].Y:= 150;

  DrawLine(img, path2, 1, Color32(clBlack), esPolygon);`

I get a two pixels width. Other line with this code :   DrawLine(img, path2, 4, Color32(clBlack), esPolygon); draws the line with 4 pixels. I try to set  img.AntiAliased:= false , but it seems doesn't affect to this. How can I draw that line with one pixel width?
![line_width](https://user-images.githubusercontent.com/60829339/227344011-2af450d9-165d-41bf-84de-056dabffd234.png)

**Comment by AngusJohnson** [comment]
Thanks for the feedback.

What's currently happening is that the renderer is drawing a line at Y = 149.5, so it's splitting the line between  Y = 149 and Y = 150.
Of course this isn't a good long term solution, but for the moment you'll need to offset your lines by 0.5 both horizontally and vertically to get the 1px width full density lines you're expecting.

Why is this offset there? Because when antialiasing is enabled, mixing perfectly solid 1px horizontal and vertical lines with other antialiased lines looks odd. But when antialias is disabled, I can see that this 0.5px offset is unhelpful and should be fixed.

**Comment by KarlheinzJansen** [comment]
![image](https://github.com/user-attachments/assets/a6313ecc-87d0-469b-87a2-5976696c0519)

Angus, 
the lines are drawn with 1 pt and zoom is done inside TImage32Panel.
I believe the double line belong to this discussion.
How can I improve it?
I am thinking of not to zoom in pixel mode and to zoom the path and draw with 1 pt line new.

**Reply by AngusJohnson** [reply]
> the lines are drawn with 1 pt and zoom is done inside TImage32Panel.
> I believe the double line belong to this discussion.
> How can I improve it?

The only way to avoid 1 pt lines being drawn over 2 lines is to disable anti-aliasing.
As I explained above, this is deliberate because doing otherwise very often produces undesireable results where axis aligned straight segments appear thinner than other (anti-aliased) line segments.

For example:
```
  img : TImage32;
  p   : TPathD;
  r   : TRectD;
begin
  img := TImage32.Create(100,100);
  r := RectD(5, 5, 95, 95);
  p := RoundRect(r, 20);
  DrawLine(img, p, 1, clBlack32, esClosed);
  img.SaveToFile('c:\temp\test2.png');
  img.Free;
```
Single pixel width line that's anti-aliased over 2 lines:
![test1](https://github.com/user-attachments/assets/7e901b4f-d462-4466-9d91-b2e320023b29)
Single pixel width line that only anti-aliased at rounded corners:
![test2](https://github.com/user-attachments/assets/3d182e04-d168-4bb2-b7c0-79870bd4f5b1)

**Reply by andersmelander** [reply]
I think there's something wrong with your AA coverage calculation; A 1 pixel wide vertical line covering two pixel columns should result in two vertical lines each with 50% alpha (~$7F000000). As far as I can tell, from your attachments, your two lines have ~62% alpha ($9F000000).
The same inaccuracy can be seen in the corners of the lower image.

I suspect that if you fix this then the smudgy artifacts in the pixel aligned corners (lower image) will be gone

**Comment by AngusJohnson** [comment]
> As far as I can tell, from your attachments, your two lines have ~62% alpha ($9F000000).

Antialiasing does indeed have a 25% weighting. 
And that's deliberate as it _mostly_ prevents background bleedthrough between 2 antialiased polgons sharing a common edge.

https://github.com/AngusJohnson/Image32/blob/1f1cec033c0c26d3a1a90c716489afbdf2a7a773/source/Img32.Draw.pas#L1361-L1362

```
  img := TImage32.Create(80,80);
  p := MakePath([0,80, 0,0, 80,0]);
  p2 := MakePath([80,0, 80,80, 0,80]);
  DrawPolygon(img, p, frNonZero, clBlack32);
  DrawPolygon(img, p2, frNonZero, clBlack32);
  img.SaveToFile('c:\temp\test2.png');
  img.Free;
```
This is **without** the 25% weighting:
![test3](https://github.com/user-attachments/assets/cba12468-19e9-48c0-b393-34ef50e47783)
This is **with** the 25% weighting:
![test4](https://github.com/user-attachments/assets/a1bdfb85-0f87-4387-b6c6-d8d13b4b1a00)

See also:
https://github.com/AngusJohnson/Image32/blob/1f1cec033c0c26d3a1a90c716489afbdf2a7a773/source/Img32.Draw.pas#L1148
https://github.com/AngusJohnson/Image32/blob/1f1cec033c0c26d3a1a90c716489afbdf2a7a773/source/Img32.Draw.pas#L1099-L1103

> I suspect that if you fix this then the smudgy artifacts in the pixel aligned corners (lower image) will be gone

I wouldn't describe antialiasing as an artifact, since the apparent blurring it produces is unavoidable.
Nevertheless, even when this blurring is lessened by 25% (by removing the weighting I mentioned above), it's still there.
![test5](https://github.com/user-attachments/assets/4879adb0-94cc-40a7-a0c1-fcb9c0654afc)

**Reply by andersmelander** [reply]
> Antialiasing does indeed have a 25% weighting.
And that's deliberate as it mostly prevents background bleedthrough between 2 antialiased polgons sharing a common edge.

I see. I also struggled with that problem when I tried to solve the stroke-and-fill problem. In theory it should be solvable by lessening rounding errors so the alpha-sum of the two edges becomes 100% but in practice that quickly becomes unfeasible. I don't have a good solution.

> I wouldn't describe antialiasing as an artifact, since the apparent blurring it produces is unavoidable.

I didn't mean AA in general. I was talking about the smudge where the straight lines transitions to an arc:
![clip](https://github.com/user-attachments/assets/d87ff363-88d4-4913-acc5-723726cd33b4)

> Nevertheless, even when this blurring is lessened by 25% (by removing the weighting I mentioned above), it's still there.

FWIW, to me that last image looks much better than either of the other two but I guess I'm biased; Here's what the same shape looks like in "that other library":

| Image32 - 25% | "other" |
| --- | --- |
| ![image32](https://private-user-images.githubusercontent.com/5280692/442559633-4879adb0-94cc-40a7-a0c1-fcb9c0654afc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcwNDA2OTgsIm5iZiI6MTc0NzA0MDM5OCwicGF0aCI6Ii81MjgwNjkyLzQ0MjU1OTYzMy00ODc5YWRiMC05NGNjLTQwYTctYTBjMS1mY2I5YzA2NTRhZmMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MTJUMDg1OTU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2MyOWMwODhjMDAyNzY3NWQ3YTg4NDE1M2FmOTEzODMwOTZmZGQzZTFmZTljMzczYjZkNTZhNDkyNDdlODE1NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.S05ZrInMpyrD_HAJ9VaMELg0Y5FujD9zBek21rV85Ds) | ![clip2](https://github.com/user-attachments/assets/ea4c2c2e-728a-4bfd-afff-1271e6006acb) |

**Reply by AngusJohnson** [reply]
> FWIW, to me that last image looks much better than either of the other two but I guess I'm biased; Here's what the same shape looks like in "that other library":

OK, I'll reconsider my current approach. 
And I may well only enhance polygon antialiasing (ie not lines too).

And I have no objection at all if you (or others) wish to mention attributes of the `Graphics32` library (or any other graphics library).

**Reply by AngusJohnson** [reply]
> Here's what the same shape looks like in "that other library"

Comparing RoundRect drawing between `Image32` and `Graphics32`, I get virtually identical rendering (when removing the 25% antialias enhancement in Image32 as discussed above). 
| Image32 - 25% | "Graphics32" |
| --- | --- |
| ![img32_roundrect_1](https://github.com/user-attachments/assets/ffa335f2-29e3-4831-9365-b4196e1147e4) | ![gr32_roundrect_1](https://github.com/user-attachments/assets/5024b92c-468e-4e05-bb50-2bc667c7cf8e) |

```
  img := TImage32.Create(100,100);
  p := RoundRect(RectD(5, 5, 95, 95), 20);
  DrawLine(img, p, 1, clBlack32, esClosed);
  img.SaveToFile('c:\temp\img32_roundrect_1.png');
  img.Free;
``` 
```
  bmp := TBitmap32.create(100,100);
  path := GR32_VectorUtils.RoundRect(FloatRect(5, 5, 95, 95), 20);
  PolylineFS(bmp, path, clBlack32, true, 1);
  bmp.SaveToFile('c:\temp\gr32_roundrect_1.bmp');
  bmp.free;
```

And in both libraries, translating the paths right & down by 0.5 px again produces virtually identical (but slightly crisper) rendering:
| Image32 - 25% | "Graphics32" |
| --- | --- |
| ![img32_roundrect_2](https://github.com/user-attachments/assets/065b2124-8498-4a27-92b2-2ba47c049a81) | ![gr32_roundrect_2](https://github.com/user-attachments/assets/4062cb8d-b966-4347-a436-fe99ac298044) |

**Reply by andersmelander** [reply]
> I get virtually identical rendering (when removing the 25% antialias enhancement in Image32 as discussed above).

Yes; That's as expected and confirms my observation.

**Comment by AngusJohnson** [comment]
https://www.angusj.com/image32/Docs/Units/Img32.Draw/Routines/DrawPolygon.htm
